runtime/lua/vim/lsp/util.lua

```diff
@@ -1361,6 +1362,45 @@ local position_sort = sort_by_key(function(v)
   return {v.start.line, v.start.character}
 end)
 
+-- Gets the zero-indexed line from the given uri.
+-- For non-file uris, we load the buffer and get the line
+-- If a loaded buffer exists, then that is used.
+-- Otherwise we get the line using libuv which is a lot faster than loading the buffer.
+--@param uri string uri of the resource to get the line from
+--@param row number zero-indexed line number
+--@return string the line at row in filename
+function M.get_line(uri, row)
+  -- load the buffer if this is not a file uri
+  -- NOTE: when does this happen?
+  if uri:sub(1, 4) ~= "file" then
+    local bufnr = vim.uri_to_bufnr(uri)
+    vim.fn.bufload(bufnr)
+    return (vim.api.nvim_buf_get_lines(bufnr, row, row + 1, false) or { "" })[1]
+  end
+
+  local filename = vim.uri_to_fname(uri)
+
+  -- use loaded buffers if available
+  if vim.fn.bufloaded(filename) == 1 then
+    local bufnr = vim.fn.bufnr(filename, false)
+    return (vim.api.nvim_buf_get_lines(bufnr, row, row + 1, false) or { "" })[1]
+  end
+
+  local fd = uv.fs_open(filename, "r", 438)
+  -- TODO: what should we do in this case?
+  if not fd then return "" end
+  local stat = uv.fs_fstat(fd)
+  local data = uv.fs_read(fd, stat.size, 0)
+  uv.fs_close(fd)
+
+  local lnum = 0
+  for line in string.gmatch(data, "([^\n]*)\n?") do
```

Comment:
I tried multiple methods to get the line and this was the simplest / fastest. Other methods I tried:
* use string.find with `plain`. was slower
* use `vim.split` slower and actually also lua
* using `vim.fn.split` was much slower
* match one regex with an endline pattern repeated for `row` times: was a bit faster, but failed for large files

If anyone has other ideas I might try, let me know :)

Reply:
Maybe give the `os.lines` version a try.

Reply:
For future reference (you probably know this), avoiding vim.* is generally a good idea for anything that might be used in a luv callback

Reply:
Did you try os.lines?

Reply:
@mjlbach yep, it was on average about two times slower.

Reply:
Perf, thanks :)